ast/ast.go

```diff
@@ -879,6 +879,7 @@ func (e *ArrayExpression) Copy() Node {
 // ObjectExpression allows the declaration of an anonymous object within a declaration.
 type ObjectExpression struct {
 	BaseNode
+	With       *Identifier `json:"with,omitempty"`
```

Comment:
We talked about this offline, but for the record:

The `with` operation must be more restricted than a binary operator on two expressions. Otherwise type inference is undecidable. Spefically the right hand side must always be an object literal. While not strictly required it is much easier to implement if the left hand side is always an identifier. So the type would look more like this:

```
type WithExpression struct {
    Left *Identifier
    Right *ObjectExpression
}
```

As such I think the current implementation of adding the details directly to the ObjectLiteral type is fine.

Reply:
I don't really care too much about the specific syntax, but I'm not sure that it is true the right hand side needs to be an object literal. I think the type system should be able to decide what attributes are inside of the object even if it is an arbitrary identifier and it will also be able to assert that it is an object itself.

I don't want to hold up this PR with that though as I would prefer we merge something and change the syntax later.